### ISSUE TITLE
optimized coordinate transform that skips proj4 failures

### ISSUE THREAD

**Issue by springmeyer** [mapbox/mapnik-vector-tile#116]
Our vector tile processor [here](https://github.com/mapbox/mapnik-vector-tile/blob/45c9717060b50460fe9440ba88406fd38209335c/src/vector_tile_processor.ipp#L1236-L1248) uses `boost::geometry::transform` and this [strategy](https://github.com/mapbox/mapnik-vector-tile/blob/master/src/vector_tile_strategy.hpp#L55-L70). The underlying strategy is sound however `boost::geometry::transform` is not the right tool for the job because:

 - We need to intelligently skip and coordinates that fail to reproject (`!prj_trans_.backward()`)
 - For the fastest conversion of geometries we need to call `vector.reserve(size)`

Therefore we need:
  - knowledge of the final size of each geometry part before starting to transform them
 -  knowledge of the previous and next coordinate in order to robustly skip invalid coordinates without creating unintended artifacts.

**Comment by springmeyer** [comment]
> We need to intelligently skip and coordinates that fail to reproject (!prj_trans_.backward())

Writing unit tests for this in mapnik-vt I hit this unhandled condition: https://github.com/mapnik/mapnik/pull/2901.

It turns out that `inf` at https://github.com/mapbox/mapnik-vector-tile/blob/0c1dd1699e954f5024525cbd085ba7778d656962/src/vector_tile_strategy.hpp#L67-L68 will flip to `-9223372036854775807` and this will trigger `ClipperLib::AddPath` to throw in `RangeTest`. So, finally we have a valid and likely way to reproduce #111

**Comment by springmeyer** [comment]
9dd2729 adds the new transformer. Due to the ability to call `std::reserve` perf is noticeably better:

```
$ ./build/Release/vtile-transform 
2192.54ms (cpu 2179.91ms)   | boost::geometry::transform
1511.22ms (cpu 1500.99ms)   | transform_visitor with reserve
```

**Comment by springmeyer** [comment]
After f6fb9d4 we are noe again handling pj_transform failures without throwing.

The goal is not to perfectly represent shapes with coordinates outside the valid bounds - to do that robustly would require clipping before reprojecting. While we did that before it would now be inconvenient. So the key here is to perfectly represent shapes with 100% valid coords and therefore to:

1) not risk aborting rendering due to invalid ones.
2) do a decent job at trying to render shapes with some invalid coords

As an example of what rendering a shape with invalid coords looks like  see the following image (which is the testcase geometry visualized in studio a few ways):

![screen shot 2015-06-10 at 7 48 36 pm](https://cloud.githubusercontent.com/assets/20300/8098841/252c85f6-0faa-11e5-9608-2aac547044d1.png)

 - Red line == the result of the new skipping transform when reprojecting from epsg:4269 -> epsg:3857. This hits `pj_transform` and skips failures resulting in a partially collapsed shape. Not pretty but expected.
 - Blue line == the result of the new skipping transform when reprojecting from epsg:4326 -> epsg:3857. In this case there are no `pj_transform` failures because proj4 is not invoked and invalid bounds are clamped by https://github.com/mapnik/mapnik/blob/352586e9d7f9276fb27c10d5449863fe67450612/include/mapnik/well_known_srs.hpp#L67-L70. So, the shape is not collapsed which is nice. Since this result is nicer we could consider clamping for more projections in the future, but that is pretty low priority in my opinion.
